ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/StatementExecutorTest.java

```diff
@@ -412,14 +455,185 @@ public void shouldEnforceReferentialIntegrity() {
+    expect(mockEngine.getPersistentQueries()).andReturn(Collections.emptySet());
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        ImmutableList.of(
+            new QueuedCommand(
+                new CommandId(Type.STREAM, name, Action.CREATE),
+                new Command(csas, Collections.emptyMap(), Collections.emptyMap())
+            ),
+            new QueuedCommand(
+                new CommandId(Type.TERMINATE, queryId.getId(), Action.EXECUTE),
+                new Command(terminate, Collections.emptyMap(), Collections.emptyMap())
+            )
+        )
+    );
+
+    // Then:
+    verify(mockParser, mockEngine, mockQuery);
+  }
 
+  @Test
+  public void shouldStartUnterminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS ";
+    final List<String> streamNames = ImmutableList.of("foo", "bar", "baz");
+    final List<PersistentQueryMetadata> queries = streamNames.stream()
+        .map(name -> mockReplayCSAS(csas + name, name, new QueryId("query-" + name)))
+        .collect(Collectors.toList());
+    expect(mockEngine.getPersistentQueries()).andStubReturn(queries);
+    for (final PersistentQueryMetadata query : queries) {
+      query.start();
+      expectLastCall();
+    }
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        streamNames.stream()
+            .map(
+                name -> new QueuedCommand(
+                    new CommandId(Type.STREAM, name, Action.CREATE),
+                    new Command(csas + name, Collections.emptyMap(), Collections.emptyMap())
+                ))
+            .collect(Collectors.toList()))
+    ;
+
+    // Then:
+    queries.forEach(EasyMock::verify);
+    verify(mockParser, mockEngine);
+  }
+
+  @Test
+  public void shouldCascade4Dot1DropStreamCommand() {
+    // Given:
+    final String drop = "DROP";
+    final DropStream mockDropStream = mockDropStream("foo");
+    expect(mockParser.parseSingleStatement(drop)).andReturn(mockDropStream);
+    expect(mockEngine.getMetaStore()).andStubReturn(mockMetaStore);
+    expect(mockMetaStore.getSource("foo"))
+        .andStubReturn(mock(StructuredDataSource.class));
+    expect(mockMetaStore.getQueriesWithSink("foo"))
+        .andStubReturn(ImmutableSet.of("query-id"));
+    expect(mockEngine.terminateQuery(eq(new QueryId("query-id")), eq(false)))
+        .andReturn(true);
+    expect(mockEngine.executeDdlStatement(drop, mockDropStream, Collections.emptyMap()))
+        .andReturn(new DdlCommandResult(true, "SUCCESS"));
+    expect(mockEngine.getPersistentQueries()).andReturn(Collections.emptySet());
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        ImmutableList.of(
+            new QueuedCommand(
+                new CommandId(Type.STREAM, "foo", Action.DROP),
+                new Command(drop, Collections.emptyMap(), null)
```

Comment:
Maybe worth pulling the `null` out into a constant to make it clear this is 'cascading' because this is null, e.g. call it `PRE_VERSION_5_NULL_ORIGNAL_PROPS` or similar.